### ISSUE TITLE
Spack arch does not recognize KNL cores.

### ISSUE THREAD

**Issue by floquet** [spack/spack#6308]
After logging into a KNL compute node on LANL's Trinitite, the command `arch` reports an **incorrect** architecture, `craype-haswell`:
```
$ spack arch
cray-CNL-haswell
```
where the **expected** answer is `cray-mic-knl`

**Process:**
1.  SSH to Trinitite login node.
2.  Grab compute node: `salloc -p knl`
3.  `module switch craype-haswell craype-mic-knl`
3.  `cd` to spack directory
4.  `. share/spack/setup-env.sh`
5.  `spack arch`

**Diagnostics:**
The `craype-mic-knl` module (20) was switched with `craype-haswell`:
```
$ module list
Currently Loaded Modulefiles:
  1) modules/3.2.10.6                               9) dmapp/7.1.1-6.0.4.0_46.2__gb8abda2.ari        17) perftools-base/6.5.2
  2) intel/17.0.4                                  10) gni-headers/5.0.11-6.0.4.0_7.2__g7136988.ari  18) PrgEnv-intel/6.0.4
  3) craype-network-aries                          11) xpmem/2.2.2-6.0.4.0_3.1__g43b0535.ari         19) cray-mpich/7.6.2
  4) craype/2.5.12                                 12) job/2.2.2-6.0.4.0_8.2__g3c644b5.ari           20) craype-mic-knl
  5) cray-libsci/17.09.1                           13) dvs/2.7_2.2.30-6.0.4.1_5.4__gd731684          21) slurm/17.02.9-SSE.0
  6) udreg/2.3.2-6.0.4.0_12.2__g2f9c3ee.ari        14) alps/6.4.1-6.0.4.0_7.2__g86d0f3d.ari          22) craype-hugepages2M
  7) ugni/6.0.14-6.0.4.0_14.1__ge7db4a2.ari        15) rca/2.2.15-6.0.4.1_13.1__g46acb0f.ari         23) Base-opts/2.4.123-6.0.4.0_10.1__g6460790.ari
  8) pmi/5.0.12                                    16) atp/2.1.1
```
```
$ lscpu
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                272
On-line CPU(s) list:   0-271
Thread(s) per core:    4
Core(s) per socket:    68
Socket(s):             1
NUMA node(s):          2
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 87
Model name:            Intel(R) Xeon Phi(TM) CPU 7250 @ 1.40GHz
Stepping:              1
CPU MHz:               1401.000
CPU max MHz:           1401.0000
CPU min MHz:           1000.0000
BogoMIPS:              2799.82
L1d cache:             32K
L1i cache:             32K
L2 cache:              1024K
NUMA node0 CPU(s):     0-271
NUMA node1 CPU(s):     
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl est tm2 ssse3 fma cx16 xtpr pdcm sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch ida arat epb pln pts dtherm fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms avx512f rdseed adx avx512pf avx512er avx512cd xsaveopt
```

spack commit 9c93c4a42bab6f319a03093d7ba7c1ddbfae9023, 2017-11-13

Related:
3477 <a href="https://github.com/spack/spack/issues/3477">Cray: Unable to target mic-knl on LANL's Trinity machine</a> (open),   
3818 <a href="https://github.com/spack/spack/issues/3818">Error in spack spec : not all arguments converted during string formatting</a> (open), 
2572 <a href="https://github.com/spack/spack/issues/2572">Spack on Cray KNL system and module detection error</a> (closed)

**Comment by KineticTheory** [comment]
I did some experimentation on trinitite and found the following behavior.

1.  When logged into a KNL node, `spack arch` returns an incorrect result as mentioned above.
2. From the front end (haswell), running `spack compiler add` generates an invalid `compilers.yaml`.  
   * There are lots of duplicate entries (3 entries for `intel@17.0.4`)
   * All of the compiler entries specify `target: any`, instead of `target: haswell` or `target: mic-knl`.
3.  I removed the spack generated `compilers.yaml` and replaced it with 

```
compilers:
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules:
    - PrgEnv-intel
    - intel/17.0.4
    - craype-mic-knl
    operating_system: CNL
    paths:
      cc: cc
      cxx: CC
      f77: ftn
      fc: ftn
    spec: intel@17.0.4
    target: mic-knl
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules:
    - PrgEnv-intel
    - intel/17.0.4
    operating_system: CNL
    paths:
      cc: cc
      cxx: CC
      f77: ftn
      fc: ftn
    spec: intel@17.0.4
    target: haswell
```

This allows `spack spec zlib %intel@17.0.4 os=CNL target=mic-knl` and ` spack spec zlib % intel@17.0.4 os=CNL target=haswell` to run.
4.  ` spack install zlib % intel@17.0.4 os=CNL target=haswell` works correctly, but  `spack install zlib %intel@17.0.4 os=CNL target=mic-knl` fails with the error:

```
==> Installing zlib
==> Error: ValueError: Can't recreate arch for spec cray-CNL-mic-knl on current arch cray-CNL-haswell; spec architecture is too different
ValueError: ValueError: Can't recreate arch for spec cray-CNL-mic-knl on current arch cray-CNL-haswell; spec architecture is too different

/usr/projects/jayenne/devs/kellyt/build/spack/lib/spack/spack/package.py:937, in architecture:
     3             environment in which this package will be built."""
     4             if not self.spec.concrete:
     5                 raise ValueError("Can only get the arch for concrete package.")
  >> 6             return spack.architecture.arch_for_spec(self.spec.architecture)
```

This error combined with the bad result from `spack arch` might be a clue to why spack is failing for @floquet.

**Comment by mamelara** [comment]
The current logic in the `cray.py` file attempts to find a default architecture by looking at which target module is loaded in a clean login session (using some shell incantation, you can see it for yourself in the `platforms/cray.py` file) It does not look at /proc/cpuinfo. I haven't been able to reproduce the error above and have been able to compile on a KNL node. 

I'm currently trying to figure out a way to tell spack which architecture we are currently on and making that be the default.

The problem at the moment is the difference between Intel's model names and codenames and creating some sort of mapping between the two. I don't even know what other chip makers do with their naming scheme so that will be another issue.

As for the compiler entries, I didn't really implement any of that but I think that `target:any` is suppose to let compilers work with any target module.

**Comment by mamelara** [comment]
`SPACK_FRONT_END` and `SPACK_BACK_END` are the environment variables. You might encounter some cross compiling bugs (which is why I really want to detect the current architecture). If you're on a haswell node and building for KNL.

**Comment by floquet** [comment]
Thanks for the helpful discussion on this week's telecon.
https://github.com/spack/spack/wiki/Telcon%3A-2017-11-30

@KineticTheory

**Comment by alalazo** [comment]
Closing as it has probably been solved by #3206. Feel free to reopen if the issue persists.
